File: Mbrit.Vegas.Web.Api/Model/WalkGameStateDto.cs
using BootFX.Common.Management;
using Mbrit.Vegas.Games;
using Mbrit.Vegas.Simulator;
using Mbrit.Vegas.Web.Api.Controllers;
using System.Diagnostics.Eventing.Reader;
using System.Text;
using System.Text.Json.Serialization;

namespace Mbrit.Vegas.Web.Api.Model
{
    public class WalkGameStateDto
    {
        [JsonPropertyName("token")]
        public string Token { get; set; }

        [JsonPropertyName("name")]
        public string Name { get; set; }

        [JsonPropertyName("piles")]
        public WalkGamePilesDto Piles { get; set; }

        [JsonPropertyName("hands")]
        public List<WalkGameHandDto> Hands { get; set; } = new List<WalkGameHandDto>();

        [JsonPropertyName("probabilitySpace")]
        public WalkGameProbabilitySpaceDto ProbabilitySpace { get; set; }

        [JsonPropertyName("hasProbabilitySpace")]
        public bool HasProbabilitySpace => this.ProbabilitySpace != null;

        [JsonPropertyName("probabilitySpaceAvailableAt")]
        public int ProbabilitySpaceAvailableAt { get; set; }

        [JsonPropertyName("spike0p5")]
        public int Spike0p5 { get; set; }

        [JsonPropertyName("spike0p5Units")]
        public int Spike0p5Units { get; set; }

        [JsonPropertyName("spike1")]
        public int Spike1 { get; set; }

        [JsonPropertyName("spike1Units")]
        public int Spike1Units { get; set; }

        [JsonPropertyName("currency")]
        public string Currency { get; set; }

        internal static WalkGameStateDto FromWalkResult(WalkResult result, IWalkGamePlayer player, HandAndUnitsPair lastBankedProfit, GameRun game)
        {
            var state = result.EndState;

            var dto = new WalkGameStateDto()
            {
                Token = game.Token,
                Name = game.Name,
                ProbabilitySpaceAvailableAt = 5,
                Spike0p5 = game.Spike0p5Win,
                Spike1 = game.Spike1Win,
                Currency = game.Currency
            };

            dto.TouchUnits(game);

            // load the hands...
            var hands = game.GetHandsOrderedByHand().ToList();
            for(var handIndex = 0; handIndex < hands.Count; handIndex++)
            {
                var hand = hands[handIndex];

                var handDto = new WalkGameHandDto()
                {
                    Index = handIndex,
                    IsDraft = false,
                    Casino = GetDefaultCasino(),
                    Game = GetDefaultGame(),
                    Action = WalkGameAction.Play,
                    Outcome = hand.Outcome,
                };

                dto.Hands.Add(handDto);
            }

            // what now?
            WalkGameActionsDto actions = null;

            // what moves do we have?
            var rounds = new ManualGameRounds(hands, game);
            var moves = player.GetNextMoves(hands.Count, rounds[0], state);

            PlayerPutInMove putInMove = null;
            PlayerPlayMove playMove = null;
            PlayerWalkMove walkMove = null;

            if (moves.Any())
            {
                foreach (var move in moves)
                {
                    if (move is PlayerPutInMove)
                        putInMove = (PlayerPutInMove)move;
                    else if (move is PlayerPlayMove)
                        playMove = (PlayerPlayMove)move;
                    else if (move is PlayerWalkMove)
                        walkMove = (PlayerWalkMove)move;
                }
            }

            if (!(game.IsWaitingOnDecision))
            {
                actions = new WalkGameActionsDto();

                var builder = new StringBuilder();

                if (putInMove != null)
                {
                    actions.CanPlay = true;
                    actions.PlayUnits = ((PlayerPutInMove)putInMove).Units;
                    actions.TouchCurrency(game.Unit);

                    builder.Append("Take ");
                    builder.Append(game.FormatUnits(actions.PlayUnits));
                    builder.Append(" (");
                    builder.Append(game.UnitsToFormattedCurrency(actions.PlayUnits));
                    builder.Append(") from your investment stack and play it.");
                }

                if(playMove != null)
                {
                    actions.CanPlay = true;
                    actions.PlayUnits = ((PlayerPlayMove)playMove).Units;
                    actions.TouchCurrency(game.Unit);

                    if (lastBankedProfit != null)
                    {
                        builder.Append("Take ");
                        builder.Append(game.FormatUnits(lastBankedProfit.Units));
                        builder.Append(" (");
                        builder.Append(game.UnitsToFormattedCurrency(actions.PlayUnits));
                        builder.Append(") from winnings and bank it, and continue to press with ");
                        builder.Append(game.FormatUnits(actions.PlayUnits));
                        builder.Append(" (");
                        builder.Append(game.UnitsToFormattedCurrency(actions.PlayUnits));
                        builder.Append(") total bet.");
                    }
                    else
                    {
                        builder.Append("Add last win to previous bet to give ");
                        builder.Append(game.FormatUnits(actions.PlayUnits));
                        builder.Append(" (");
                        builder.Append(game.UnitsToFormattedCurrency(actions.PlayUnits));
                        builder.Append(") total bet.");
                    }
                }

                actions.Instructions = builder.ToString();

                if (walkMove != null)
                    actions.CanWalk = true;
            }

            // are we waiting on a decision? if so, the amount in play have to change the piles...
            if (game.IsWaitingOnDecision)
            {
                var toMove = 0;

                if (game.Action == WalkGameAction.Play && putInMove != null)
                    toMove = game.ActionUnits;

                if(toMove > 0)
                {
                    for (var index = 0; index < toMove; index++)
                        state.PutIn();
                }
            }

            // add the hand...
            var draftHand = new WalkGameHandDto()
            {
                Index = hands.Count,
                IsDraft = true,
                NeedsAnswer = game.IsWaitingOnDecision,
                Casino = GetDefaultCasino(),
                Game = GetDefaultGame(),
                //HasSeenSpike0p5 = state.HasSeenSpike0p5,
                //IsAtOrOverSpike0p5 = state.IsAtOrOverSpike0p5,
                //HasSeenSpike1 = state.HasSeenSpike1,
                //IsAtOrOverSpike1 = state.IsAtOrOverSpike1,
                //PilesBefore = WalkGamePilesDto.FromWalkState(state, game),
                //ProbabilitySpaceAvailableAt = 5,
                Actions = actions,
                Outcome = WinLoseDrawType.Win       // supersition...
            };

            dto.Hands.Add(draftHand);

            dto.Piles = WalkGamePilesDto.FromWalkState(state, game);

            // return...
            return dto;
        }

        private void TouchUnits(GameRun game)
        {
            this.Spike0p5Units = game.CurrencyToUnits(this.Spike0p5);
            this.Spike1Units = game.CurrencyToUnits(this.Spike1);
        }

        private static CasinoDto GetDefaultCasino()
        {
            return new CasinoDto()
            {
                Name = "Flamingo",
                Location = new LocationDto()
                {
                    Name = "Las Vegas Strip"
                }
            };
        }

        private static GameItemDto GetDefaultGame()
        {
            return new GameItemDto()
            {
                Name = "Baccarat Banker Bet",
                HouseEdge = 0.016M,
            };
        }
    }
}

File: Mbrit.Vegas.App/mbrit_vegas/lib/models/walk_game_state_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'walk_game_projection_dto.dart';

part 'walk_game_state_dto.g.dart';

@JsonSerializable()
class WalkGameStateDto {
  @JsonKey(name: 'token')
  final String token;

  @JsonKey(name: 'piles')
  final WalkGamePilesDto piles;

  @JsonKey(name: 'hands')
  final List<WalkGameHandDto> hands;

  @JsonKey(name: 'probabilitySpace')
  final WalkGameProbabilitySpaceDto? probabilitySpace;

  @JsonKey(name: 'hasProbabilitySpace')
  final bool? hasProbabilitySpace;

  @JsonKey(name: 'probabilitySpaceAvailableAt')
  final int? probabilitySpaceAvailableAt;

  @JsonKey(name: 'spike0p5')
  final int? spike0p5;

  @JsonKey(name: 'spike0p5Units')
  final int? spike0p5Units;

  @JsonKey(name: 'spike1')
  final int? spike1;

  @JsonKey(name: 'spike1Units')
  final int? spike1Units;

  const WalkGameStateDto({
    required this.token,
    required this.piles,
    required this.hands,
    this.probabilitySpace,
    this.hasProbabilitySpace,
    this.probabilitySpaceAvailableAt,
    this.spike0p5,
    this.spike0p5Units,
    this.spike1,
    this.spike1Units,
  });

  factory WalkGameStateDto.fromJson(Map<String, dynamic> json) =>
      _$WalkGameStateDtoFromJson(json);

  Map<String, dynamic> toJson() => _$WalkGameStateDtoToJson(this);
}

@JsonSerializable()
class WalkGameHandDto {
  @JsonKey(name: 'index')
  final int index;

  @JsonKey(name: 'isDraft')
  final bool isDraft;

  @JsonKey(name: 'casino')
  final CasinoDto casino;

  @JsonKey(name: 'game')
  final GameItemDto game;

  /*
  @JsonKey(name: 'pilesBefore')
  final WalkGamePilesDto pilesBefore;

  @JsonKey(name: 'hasSeenSpike0p5')
  final bool hasSeenSpike0p5;

  @JsonKey(name: 'isOverSpike0p5')
  final bool isOverSpike0p5;

  @JsonKey(name: 'hasSeenSpike1')
  final bool hasSeenSpike1;

  @JsonKey(name: 'isOverSpike1')
  final bool isOverSpike1;
  */

  @JsonKey(name: 'actions')
  final WalkGameActionsDto? actions;

  @JsonKey(name: 'action')
  final WalkGameAction? action;

  @JsonKey(name: 'outcome')
  final String? outcome;

  @JsonKey(name: 'needsAnswer')
  final bool needsAnswer;

  const WalkGameHandDto({
    required this.index,
    required this.isDraft,
    required this.casino,
    required this.game,
    /*
    required this.pilesBefore,
    required this.hasSeenSpike0p5,
    required this.isOverSpike0p5,
    required this.hasSeenSpike1,
    required this.isOverSpike1,
    */
    this.actions,
    this.action,
    this.outcome,
    required this.needsAnswer,
  });

  bool get hasActions => actions != null;

  factory WalkGameHandDto.fromJson(Map<String, dynamic> json) =>
      _$WalkGameHandDtoFromJson(json);

  Map<String, dynamic> toJson() => _$WalkGameHandDtoToJson(this);
}

@JsonSerializable()
class WalkGamePilesDto {
  @JsonKey(name: 'investable')
  final int investable;

  @JsonKey(name: 'investableUnits')
  final int investableUnits;

  @JsonKey(name: 'inPlay')
  final int inPlay;

  @JsonKey(name: 'inPlayUnits')
  final int inPlayUnits;

  @JsonKey(name: 'banked')
  final int banked;

  @JsonKey(name: 'bankedUnits')
  final int bankedUnits;

  @JsonKey(name: 'cashInHand')
  final int cashInHand;

  @JsonKey(name: 'cashInHandUnits')
  final int cashInHandUnits;

  @JsonKey(name: 'profit')
  final int profit;

  @JsonKey(name: 'profitUnits')
  final int profitUnits;

  const WalkGamePilesDto({
    required this.investable,
    required this.investableUnits,
    required this.inPlay,
    required this.inPlayUnits,
    required this.banked,
    required this.bankedUnits,
    required this.cashInHand,
    required this.cashInHandUnits,
    required this.profit,
    required this.profitUnits,
  });

  factory WalkGamePilesDto.fromJson(Map<String, dynamic> json) =>
      _$WalkGamePilesDtoFromJson(json);

  Map<String, dynamic> toJson() => _$WalkGamePilesDtoToJson(this);
}

@JsonSerializable()
class GameItemDto {
  @JsonKey(name: 'name')
  final String name;

  @JsonKey(name: 'houseEdge')
  final double houseEdge;

  @JsonKey(name: 'notes')
  final String? notes;

  const GameItemDto({required this.name, required this.houseEdge, this.notes});

  factory GameItemDto.fromJson(Map<String, dynamic> json) =>
      _$GameItemDtoFromJson(json);

  Map<String, dynamic> toJson() => _$GameItemDtoToJson(this);
}

@JsonSerializable()
class CasinoDto {
  @JsonKey(name: 'name')
  final String name;

  @JsonKey(name: 'location')
  final LocationDto location;

  const CasinoDto({required this.name, required this.location});

  factory CasinoDto.fromJson(Map<String, dynamic> json) =>
      _$CasinoDtoFromJson(json);

  Map<String, dynamic> toJson() => _$CasinoDtoToJson(this);
}

@JsonSerializable()
class LocationDto {
  @JsonKey(name: 'name')
  final String name;

  const LocationDto({required this.name});

  factory LocationDto.fromJson(Map<String, dynamic> json) =>
      _$LocationDtoFromJson(json);

  Map<String, dynamic> toJson() => _$LocationDtoToJson(this);
}

@JsonSerializable()
class WalkGameActionsDto {
  @JsonKey(name: 'instructions')
  final String? instructions;

  @JsonKey(name: 'canPlay')
  final bool? canPlay;

  @JsonKey(name: 'play')
  final int? play;

  @JsonKey(name: 'playUnits')
  final int? playUnits;

  @JsonKey(name: 'canHailMary')
  final bool? canHailMary;

  @JsonKey(name: 'hailMary')
  final int? hailMary;

  @JsonKey(name: 'hailMaryUnits')
  final int? hailMaryUnits;

  @JsonKey(name: 'canWalk')
  final bool? canWalk;

  const WalkGameActionsDto({
    this.instructions,
    this.canPlay,
    this.play,
    this.playUnits,
    this.canHailMary,
    this.hailMary,
    this.hailMaryUnits,
    this.canWalk,
  });

  factory WalkGameActionsDto.fromJson(Map<String, dynamic> json) =>
      _$WalkGameActionsDtoFromJson(json);

  Map<String, dynamic> toJson() => _$WalkGameActionsDtoToJson(this);
}

enum WalkGameAction {
  @JsonValue('None')
  none,
  @JsonValue('Play')
  play,
  @JsonValue('HailMary')
  hailMary,
  @JsonValue('Walk')
  walk,
}

enum WinLoseDrawType {
  @JsonValue('Lose')
  lose,
  @JsonValue('Win')
  win,
}

@JsonSerializable()
class WalkGameProbabilitySpaceDto extends WalkOutcomesBucketDto {
  const WalkGameProbabilitySpaceDto({
    required super.majorBustPercentage,
    required super.minorBustPercentage,
    required super.evensPercentage,
    required super.spike0p5Percentage,
    required super.spike1Percentage,
    required super.spike1PlusPercentage,
    required super.averageProfitWhenWon,
    required super.averageCoinIn,
  });

  factory WalkGameProbabilitySpaceDto.fromJson(Map<String, dynamic> json) =>
      _$WalkGameProbabilitySpaceDtoFromJson(json);

  @override
  Map<String, dynamic> toJson() => _$WalkGameProbabilitySpaceDtoToJson(this);
}
